addons/stock/models/stock_rule.py

```diff
@@ -255,7 +255,9 @@ def _run_pull(self, procurements):
                 if float_compare(qty_needed, 0, precision_rounding=procurement.product_id.uom_id.rounding) <= 0:
                     procure_method = 'make_to_order'
                     for move in procurement.values.get('group_id', self.env['procurement.group']).stock_move_ids:
-                        if move.rule_id == rule and float_compare(move.product_uom_qty, 0, precision_rounding=move.product_uom.rounding) > 0:
+                        if move.rule_id == rule and float_compare(move.product_uom_qty, 0, precision_rounding=move.product_uom.rounding) > 0 and \
+                           (procurement.values.get('sale_line_id') and move.sale_line_id.id == procurement.values['sale_line_id'] or
+                           procurement.values.get('sale_line_id') is None and move.product_id == procurement.product_id):
```

Comment:
Better but still too complicated to read:
- if you place all the conditions inside one big parenthesis, you can go to next line without needing a backslash
- your `or` statement is not clear. What are its possibilities? Try to isolate it inside its own parenthesis if possible.

```py
if (condition_1
        and condition_2
        and (condition_3 or condition_4)
        and condition_5):
    do_stuff()
```

Reply:
also please update your commit message with why you use those specific conditions